org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/embedder/MavenExecutionContext.java

```diff
@@ -132,25 +153,30 @@ public MavenExecutionRequest getExecutionRequest() throws CoreException {
   protected MavenExecutionRequest newExecutionRequest() throws CoreException {
     MavenExecutionRequest request = null;
     Deque<MavenExecutionContext> stack = threadLocal.get();
+    File parentMultiModuleProjectDirectory = null;
```

Comment:
I'm not sure anymore why I added that part. Maybe for the case of a nested `.mvn` folder?
I have to check it.

Reply:
It might be that there are still places that do use the "global" maven where the should have used a project scoped (or basedir scoped) execution instead. I think I already have catched some but not all, as I mostly focused o getting to a working state again. SO if you find out where this was required we can take a look if there are better ways to solve it.

In general one can say there are three cases:

1. whenever you can get an IMavenFacade --> Use it for execution as you most probably want project scope execution
2. whenever you do anything "external" (e.g. as we do in inital project setup or when reading external-to-the-workspace poms) create a basedir scoped execution (IMavenExecution.of(...))
3. when there is the rare case you just want do something "global" that is not bound to a project or directory (e.g. maven targets just want fetch artifacts using the global maven config) you can use IMaven as the global context

Reply:
That makes sense.
This list should definitely be documented somewhere.

Reply:
> I'm not sure anymore why I added that part. Maybe for the case of a nested `.mvn` folder? I have to check it.

I checked it and that is necessary when a execution request is created for a nested context, i.e. when a 'surrounding' context is already on the stack. Then the parent is just copied and often has a null `basedir/multiModuleProjectDir`.

This happens for example when importing projects. Then within the execution of a global Maven context a context with basedir is created. But since there is already a context on the stack, that parent context's request is copied, but that does not have basedir specific settings.

But maybe the check should be changed from `if(!Objects.equals(multiModuleProjectDirectory, parentMultiModuleProjectDirectory))` to `if(parentMultiModuleProjectDirectory==null)`.

Reply:
I'm not sure if nesting is the right approach then here. Nesting only is useful if you want to fully inherit the context of the caller (e.g. you perform a generic method) but I'm not sure if it is actually useful at all here, so the question is why there is a surrounding context at all and if it could not be replaced with one of the other alternatives (or most probably must!).

Reply:
I reduced the number of cases where this happened by also copying the `multiModuleProjectDirectory` from the original to the copy when a parent is present. For whatever reason `DefaultMavenExecutionRequest.copy()` does not copy that.

However that case still happens when for example opening a closed project that has a `.mvn/maven.config`.
And as long as the thread-local context stack principle is used, I think it actually makes sense in such cases to copy the outer context and then loads again the properties, etc. from the specific `.mvn/maven.config` file for the inner context.
And such cases happen for example on project-refresh that is started with the global context.

Of course we could refactor the `ProjectRegistryRefreshJob` to only use basedir aware contexts, maybe even project specific contexts and should check for other locations where that could be a problem, that logic should remain.
And I don't know if it can happen, that within the execution of some basedir-aware context another context with a deeper nested basedir can be created in corresponding scenarios.
If that can happen we should keep that otherwise it could be hard for users to do it right.

Reply:
> For whatever reason DefaultMavenExecutionRequest.copy() does not copy that.

Most probably it was forgotten and you probably want to propose a fix to maven.

> And as long as the thread-local context stack principle is used, I think it actually makes sense in such cases to copy the outer context and then loads again the properties, etc. from the specific .mvn/maven.config file for the inner context.

To be honest I'm not convinced. Mixing global with project context seems not right, its not only that directory, its also the container lookup and the class worlds are different.

> And such cases happen for example on project-refresh that is started with the global context.

Then this must be fixed (e.g. with a separate PR) similar to how the initial reading of projects is performed by grouping the projects and container appropriately.

> Of course we could refactor the ProjectRegistryRefreshJob

Actually this should be done. I think the intial idea here is/was that because refresh is maybe called multiple times, we can reuse the session and thus using a stacked execution. This won't work well with the new concept and one would either need to accepts that sessions are created multiple times (I think that's really fast) and let the manager do the grouping instead.